fix(zone.js): add conditional exports to zone.js package

alan-agius4 opened the pull request:
This is needed to better support native ESM modules and avoid the otherwise necessary deep imports like `zone.js/fesm2015/zone-node.js` due to disallowed directory imports.

alan-agius4 committed:
fix(zone.js): add conditional exports to zone.js package

This is needed to better support native ESM modules and avoid the otherwise necessary deep imports like `zone.js/fesm2015/zone-node.js` due to disallowed directory imports.

alan-agius4 committed:
fixup! fix(zone.js): add conditional exports to zone.js package

clydin commented on packages/zone.js/package.json:
```diff
@@ -42,5 +42,314 @@
+    },
+    ".": {
+      "typings": "./zone.d.ts",
+      "import": "./fesm2015/zone.js",
```
nit: The `import` condition is not needed if the same file is also the `default` condition.
This change can also reduce the overall size of the `package.json` somewhat.

clydin commented on packages/zone.js/package.json:
```diff
@@ -42,5 +42,314 @@
+      "require": "./bundles/zone-mix.umd.js",
+      "default": "./fesm2015/zone-mix.js"
+    },
+    "./plugins/async-test": {
+      "import": "./fesm2015/async-test.js",
+      "require": "./bundles/async-test.umd.js",
+      "default": "./fesm2015/async-test.js"
+    },
+    "./plugins/async-test.min": {
+      "import": "./fesm2015/async-test.min.js",
+      "require": "./bundles/async-test.umd.min.js",
+      "default": "./fesm2015/async-test.min.js"
+    },
```
```suggestion
    "./plugins/*.min": {
      "import": "./fesm2015/*.min.js",
      "require": "./bundles/*.umd.min.js",
      "default": "./fesm2015/*.min.js"
    },
    "./plugins/*": {
      "import": "./fesm2015/*.js",
      "require": "./bundles/*.umd.js",
      "default": "./fesm2015/*.js"
    },
```
The subpath patterns should allow condensing the individual plugin entries into just two.

JiaLiPassion reviewed (approved):
LGTM!

alan-agius4 committed:
fixup! fix(zone.js): add conditional exports to zone.js package

alan-agius4 committed:
ci: force new yarn-lock file to invalidate cache

This commit update the lock file to force a new `node_modules` cache as this causes the patch to fail due to GitHub caching and at present, there is also no way to provide a key to invalidate the cache.

clydin reviewed (approved):
LGTM

devversion reviewed (approved):
Nice, only thing that will be a little slower is having to build the full zone npm package for benchmarks- but I think that is reasonable

AviVahl commented on packages/zone.js/package.json:
```diff
@@ -42,5 +42,41 @@
+      "default": "./package.json"
+    },
+    ".": {
+      "typings": "./zone.d.ts",
```
this must say "`types"` for typescript to pick it up when project uses `"moduleResolution"` that respects `"exports"` (e.g. `moduleResolution": "bundler"`).

alan-agius4 replied:
Thanks for bringing this up. Here's a fix https://github.com/angular/angular/pull/51737
